types/pmx/tcr/uplink.js

```diff
@@ -0,0 +1,330 @@
+* @param {Date} input.recvTime - The uplink message time recorded by the LoRaWAN network server
+* @returns {DecodedUplink} The decoded object
+*/
+function decodeUplink(input) {
+
+}
+
+
+
+function makeError(desc) {
+  return {
+    error: { error: desc },
+  };
+}
+
+function deleteUnusedKeys(data) {
+  let keysRetained = false;
+  Object.keys(data).forEach((key) => {
+    if (data[key] === undefined || isNaN(data[key])) {
+      delete data[key];
+    } else {
+      keysRetained = true;
+    }
+  });
+  return keysRetained;
+}
+
+function consume(event) {
+  const bytes = Hex.hexToBytes(event.data.payloadHex);
+  const port = event.data.port;
+
+  var counts = {};
+  var lifecycle = {};
+  var id = {};
+
+  // it's a Device ID Payload V2 (PMX Firmware for TCR)
+  if (port == 190 && bytes[0] == 0xd2) {
+    let topic = "id";
+    let data = {};
+
+    id = d2_decoder(bytes, port);
+
+    data.type = id.typestr;
+    data.licence = id.fu_level;
+    data.speedclass = id.speedclass;
+    data.firmware = id.fw_version;
+    data.sbx = id.sbx_version
+
+    emit("sample", { data, topic });
+  }
+
+  // it's a Counter Payload V2 (PMX Firmware for TCR)
+  if (bytes[0] == 0xa2) {
+    let topic = "default";
+    let data = {};
+    let time = new Date();
+
+    let cnt = a2_decoder(bytes, port);  // decode counter payload
+
+    switch (port) {
+      case 13:
+        topic = "udc";
+        data.LTRCounter = cnt.udc_ltr_cnt;
+        data.LTRSpeed = cnt.udc_ltr_spd;
+        data.RTLCounter = cnt.udc_rtl_cnt;
+        data.RTLSpeed = cnt.udc_rtl_spd;
+        time.setHours(cnt.udc_hours);
+        time.setMinutes(cnt.udc_minutes);
+        break;
+
+      case 14:
+        topic = "cat1";
+        data.LTRCounter = cnt.cat1_ltr_cnt;
+        data.LTRSpeed = cnt.cat1_ltr_spd;
+        data.RTLCounter = cnt.cat1_rtl_cnt;
+        data.RTLSpeed = cnt.cat1_rtl_spd;
+        time.setHours(cnt.cat1_hours);
+        time.setMinutes(cnt.cat1_minutes);
+        break;
+
+      case 15:
+        topic = "cat2";
+        data.LTRCounter = cnt.cat2_ltr_cnt;
+        data.LTRSpeed = cnt.cat2_ltr_spd;
+        data.RTLCounter = cnt.cat2_rtl_cnt;
+        data.RTLSpeed = cnt.cat2_rtl_spd;
+        time.setHours(cnt.cat2_hours);
+        time.setMinutes(cnt.cat2_minutes);
+        break;
+
+      case 16:
+        topic = "cat3";
+        data.LTRCounter = cnt.cat3_ltr_cnt;
+        data.LTRSpeed = cnt.cat3_ltr_spd;
+        data.RTLCounter = cnt.cat3_rtl_cnt;
+        data.RTLSpeed = cnt.cat3_rtl_spd;
+        time.setHours(cnt.cat3_hours);
+        time.setMinutes(cnt.cat3_minutes);
+        break;
+
+      case 17:
+        topic = "cat4";
+        data.LTRCounter = cnt.cat4_ltr_cnt;
+        data.LTRSpeed = cnt.cat4_ltr_spd;
+        data.RTLCounter = cnt.cat4_rtl_cnt;
+        data.RTLSpeed = cnt.cat4_rtl_spd;
+        time.setHours(cnt.cat4_hours);
+        time.setMinutes(cnt.cat4_minutes);
+        break;
+    }
+    emit("sample", { data, topic });
```

Comment:
in order to emit a sample at a timestamp, change this to `emit("sample", { data, topic, timestamp });`